design/lvm/mountable_snapshot.md

```diff
@@ -0,0 +1,191 @@
+- In the [agent.go](../../pkg/driver/agent.go) the `NodePublishVolume` function for the `node` type needs to be changed such that it checks whether the `volName` is a snapshot and if so mount the snapshot to the specified location.
+  - This also needs to change the write access to `rw` by using the lvchange command when the PersistentVolumeClaim specified the AccessMode as ReadWrite. Note that we do not need to change it back to read only, since we can limit the permissions of future mounts to read only by using the `MountOptions`. 
+    - Alternatively we could make the Snapshots writeable by default
+- In the [agent.go](../../pkg/driver/agent.go) the `NodeUnpublishVolume` function for the `node` type needs to be changed such that it unmounts the snapshot.
+
+
+## Test Plan
+
+- Create the PersistentVolumeClaim for the snapshot and verify that it is created successfully and that it can successfully be bound
+- Create the PersistentVolumeClaim for the thick snapshot with a storage size larger than the snapSize but less than the size of the original volume and verify that the storage size has increased
+- Create the PersistentVolumeClaim for the thick snapshot with a storage size larger than the size of the original volume and verify that the size has been set to exactly the size of the original volume
+- Create the PersistentVolumeClaim for the thin snapshot with storage size larger than that of the snapshot and verify that it has not been resized
+- Mount a PersistentVolumeClaim of a VolumeSnapshot to a pod and verify that it is mounted successfully
+- Delete the VolumeSnapshot and verify that all PersistentVolumeClaims that reference it have been deleted
+- Delete the VolumeSnapshot and verify that, once all the pods that have mounted a PersistentVolumeClaim with as dataSource the VolumeSnapshot have been deleted, the Snapshot is deleted
+- Mount the PersistentVolumeClaim to a pod as Read and validate that the filesystem is read only
+- Mount the PersistentVolumeClaim to a pod as ReadWrite and validate that the filesystem is writable
+- Verify the original volume is working fine after mounting the snapshot and that any changes to the snapshot are not propagated to the original volume and vice versa
+- Verify that creating a PersistentVolumeClaim for a non existing VolumeSnapshot fails
+
+## Graduation Criteria
+
+All testcases mentioned in [Test Plan](#test-plan) section need to be automated
+
+## Drawbacks
+
+- As far as I understand it, normally when creating a PersistentVolumeClaim with a snapshot as datasource it would clone the Snapshot into an actual volume, so we'd be diverging from this behaviour.
+  - We could use an annotation to specify the behaviour that we are implementing here so that the clone behaviour can be added at a later time?
```

Comment:
can you mention the annotation and also how we are going to use that?

Reply:
I've added some additional details to the document. Basically we annotate the PVC to indicate whether it needs to clone the snapshot or mount it.